### ISSUE TITLE
Pylint erroneously reports too-many-function-args for typing.NamedTuple classes

### ISSUE THREAD

**Issue by eddie-dunn** [PyCQA/pylint#1789]
### Steps to reproduce

Create a Python 3.6.1+ [NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) class:

```python
"""Docstring"""
from typing import NamedTuple


class Foobar(NamedTuple):  # pylint: disable=too-few-public-methods
    """Docstring"""
    boolean: bool
    msg: str = ''
    data: dict = {}


print(Foobar(True, 'hello', {'a': 1}))
```

### Current behavior

```
$ pylint foob.py
Using config file /home/user/.config/pylintrc
************* Module foob
E: 12, 6: Too many positional arguments for constructor call (too-many-function-args)

------------------------------------------------------------------
Your code has been rated at 1.67/10 (previous run: 1.67/10, +0.00)
```

### Expected behavior
```
$ pylint foob.py
Using config file /home/user/.config/pylintrc

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

### pylint --version output
```
$ pylint --version
Using config file /home/user/.config/pylintrc
pylint 1.8.1, 
astroid 1.6.0
Python 3.6.2 (default, Sep 22 2017, 12:45:49) 
[GCC 6.3.0 20170516]
```

**Comment by PCManticore** [comment]
I tried reproducing this using Python 3.6.2, but seems to work fine. Have to check if 3.6.1 has something different that might trigger this.

**Comment by PCManticore** [comment]
Can't reproduce it on 3.6.1 as well. The only thing I'm getting is:
```
************* Module a
C:  1, 0: Missing module docstring (missing-docstring)
```

**Comment by gangefors** [comment]
I am not able to reproduce it on my local machine, but I manage to reproduce it in a docker container. So I do think there is something strange going on here.

```
docker run -it -v `pwd`/foob.py:/tmp/foob.py -w /tmp python:3.6 bash 
root@204ee5cd07e0:/tmp# pip install pylint
Collecting pylint
  Downloading pylint-1.8.1-py2.py3-none-any.whl (683kB)
    100% |████████████████████████████████| 686kB 2.1MB/s 
Collecting mccabe (from pylint)
  Downloading mccabe-0.6.1-py2.py3-none-any.whl
Collecting six (from pylint)
  Downloading six-1.11.0-py2.py3-none-any.whl
Collecting isort>=4.2.5 (from pylint)
  Downloading isort-4.2.15-py2.py3-none-any.whl (43kB)
    100% |████████████████████████████████| 51kB 2.0MB/s 
Collecting astroid<2.0 (from pylint)
  Downloading astroid-1.6.0-py2.py3-none-any.whl (288kB)
    100% |████████████████████████████████| 296kB 3.6MB/s 
Collecting lazy-object-proxy (from astroid<2.0->pylint)
  Downloading lazy_object_proxy-1.3.1-cp36-cp36m-manylinux1_x86_64.whl (55kB)
    100% |████████████████████████████████| 61kB 6.0MB/s 
Collecting wrapt (from astroid<2.0->pylint)
  Downloading wrapt-1.10.11.tar.gz
Building wheels for collected packages: wrapt
  Running setup.py bdist_wheel for wrapt ... done
  Stored in directory: /root/.cache/pip/wheels/56/e1/0f/f7ccf1ed8ceaabccc2a93ce0481f73e589814cbbc439291345
Successfully built wrapt
Installing collected packages: mccabe, six, isort, lazy-object-proxy, wrapt, astroid, pylint
Successfully installed astroid-1.6.0 isort-4.2.15 lazy-object-proxy-1.3.1 mccabe-0.6.1 pylint-1.8.1 six-1.11.0 wrapt-1.10.11
root@204ee5cd07e0:/tmp# pylint foob.py 
No config file found, using default configuration
************* Module foob
E: 12, 6: Too many positional arguments for constructor call (too-many-function-args)

-----------------------------------
Your code has been rated at 1.67/10
```

**Comment by PCManticore** [comment]
I tried the same thing and here's what I'm getting:

```
root@3462e4ac22ad:/tmp# cat a.py
"""Docstring"""
from typing import NamedTuple


class Foobar(NamedTuple):  # pylint: disable=too-few-public-methods
    """Docstring"""
    boolean: bool
    msg: str = ''
    data: dict = {}


print(Foobar(True, 'hello', {'a': 1}))
root@3462e4ac22ad:/tmp# pylint a.py
No config file found, using default configuration

------------------------------------
Your code has been rated at 10.00/10

root@3462e4ac22ad:/tmp# pylint --version
No config file found, using default configuration
pylint 1.8.1, 
astroid 1.6.0
Python 3.6.2 (default, Sep 13 2017, 14:26:54) 
[GCC 4.9.2]
```

**Comment by gangefors** [comment]
This is extremely annoying. I have two colleagues that get the reported error consistently both on their machines and in a docker container. I can only reproduce it in a docker container, not on my local machine. And you can't reproduce it at all.

What is going on here?

**Comment by PCManticore** [comment]
Not sure, but If you can reproduce this consistently it makes sense to reopen it. It might also be related to various dependencies, we'll have to take a look since I have no clue why is it happening

**Comment by gangefors** [comment]
We will take a closer look at it and compare environments on the different machines to see if we can come to a conclusion to why this is happening. We have now observed different behavior in our Jenkins environment where we run tests in a docker container, one slave shows the error when another one does not.

**Comment by OJFord** [comment]
I also have this issue, not in a container, (darwin native) with python 3.6.4.

**Comment by kstrauser** [comment]
Another data point for Python 3.6.4 via Homebrew on macOS 10.13.3.

**Comment by PCManticore** [comment]
Tried again with the Python versions mentioned so far and I still can't reproduce this bug using the master branch for both pylint and astroid. If you can, please test with these two versions and if it's not working for you, let me know. Closing the issue in the meantime.

**Comment by OJFord** [comment]
Not working in pylint v1.8.4; fixed in v1.9.0.

**Comment by nicktimko** [comment]
Depending on the name of the imported `NamedTuple` I get different errors... It seems like it's kinda aware of NamedTuple being something special, but not quite, because it won't error on too-few arguments. It's a black box to me right now, so I'm kinda fumbling around looking for what works and what doesn't.

`namedtuplelint.py`
```python
import typing
t = typing
from typing import NamedTuple

class NT(NamedTuple):
    a: str
    b: str
    c: str

class TNT(t.NamedTuple):
    """pylint doesn't like this guy..."""
    a: str
    b: str
    c: str

class TypingNT(typing.NamedTuple):
    a: str
    b: str
    c: str

NT("a", "b")            # no err, should be E____: not enough args?
NT("a", "b", "c")
NT("a", "b", "c", "d")  # no err, should be E1121: too many args

TNT("a", "b")            # no err, should be not enough args?
TNT("a", "b", "c")       # E1121 (false-positive)
TNT("a", "b", "c", "d")  # E1121 (expected)

TypingNT("a", "b")            # no err, should be E____: not enough args?
TypingNT("a", "b", "c")
TypingNT("a", "b", "c", "d")  # no err, should be E1121: too many args
```

`Dockerfile`
```docker
FROM python:3.7-alpine
RUN pip3 install pylint \
    && pylint --version
COPY namedtuplelint.py /app/namedtuplelint.py
CMD ["pylint", "-E", "/app/namedtuplelint.py"]
```

Running...
```
> sh -c "docker build --no-cache -t namedtuplelint . && docker run -it namedtuplelint"                                            
Sending build context to Docker daemon    212kB                                                                                                                 
Step 1/4 : FROM python:3.7-alpine                                                                                                                               
 ---> aadc3feb2b19                                                                                                                                              
Step 2/4 : RUN pip3 install pylint     && pylint --version                                                                                                      
 ---> Running in 35188e9548ec                                                                                                                                   
Collecting pylint                                                                                                                                               
  Downloading https://files.pythonhosted.org/packages/a5/06/ecef826f319055e6b231716730d7f9047dd7524ffda224b521d989f085b6/pylint-2.2.2-py3-none-any.whl (750kB)  
Collecting isort>=4.2.5 (from pylint)                                                                                                                           
  Downloading https://files.pythonhosted.org/packages/1f/2c/22eee714d7199ae0464beda6ad5fedec8fee6a2f7ffd1e8f1840928fe318/isort-4.3.4-py3-none-any.whl (45kB)    
Collecting mccabe (from pylint)                                                                                                                                 
  Downloading https://files.pythonhosted.org/packages/87/89/479dc97e18549e21354893e4ee4ef36db1d237534982482c3681ee6e7b57/mccabe-0.6.1-py2.py3-none-any.whl      
Collecting astroid>=2.0.0 (from pylint)                                                                                                                         
  Downloading https://files.pythonhosted.org/packages/fc/53/8809bc008bad0300897281a7b320b286dc0e84e836396c0cff6279841e8a/astroid-2.1.0-py3-none-any.whl (176kB) 
Collecting wrapt (from astroid>=2.0.0->pylint)                                                                                                                  
  Downloading https://files.pythonhosted.org/packages/a0/47/66897906448185fcb77fc3c2b1bc20ed0ecca81a0f2f88eda3fc5a34fc3d/wrapt-1.10.11.tar.gz                   
Collecting lazy-object-proxy (from astroid>=2.0.0->pylint)                                                                                                      
  Downloading https://files.pythonhosted.org/packages/55/08/23c0753599bdec1aec273e322f277c4e875150325f565017f6280549f554/lazy-object-proxy-1.3.1.tar.gz         
Collecting six (from astroid>=2.0.0->pylint)                                                                                                                    
  Downloading https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl        
Building wheels for collected packages: wrapt, lazy-object-proxy                                                                                                
  Running setup.py bdist_wheel for wrapt: started                                                                                                               
  Running setup.py bdist_wheel for wrapt: finished with status 'done'                                                                                           
  Stored in directory: /root/.cache/pip/wheels/48/5d/04/22361a593e70d23b1f7746d932802efe1f0e523376a74f321e                                                      
  Running setup.py bdist_wheel for lazy-object-proxy: started                                                                                                   
  Running setup.py bdist_wheel for lazy-object-proxy: finished with status 'done'                                                                               
  Stored in directory: /root/.cache/pip/wheels/a0/63/e2/6d93295282cb35b53b14b50b602c76dfb04471e21b31d8ad7b                                                      
Successfully built wrapt lazy-object-proxy                                                                                                                      
Installing collected packages: isort, mccabe, wrapt, lazy-object-proxy, six, astroid, pylint                                                                    
Successfully installed astroid-2.1.0 isort-4.3.4 lazy-object-proxy-1.3.1 mccabe-0.6.1 pylint-2.2.2 six-1.12.0 wrapt-1.10.11                                     
pylint 2.2.2                                                                                                                                                    
astroid 2.1.0                                                                                                                                                   
Python 3.7.1 (default, Nov 16 2018, 06:29:01)                                                                                                                   
[GCC 6.4.0]                                                                                                                                                     
Removing intermediate container 35188e9548ec                                                                                                                    
 ---> c5ddb623dd1f                                                                                                                                              
Step 3/4 : COPY namedtuplelint.py /app/namedtuplelint.py                                                                                                        
 ---> 6e6a935b69cc                                                                                                                                              
Step 4/4 : CMD ["pylint", "-E", "/app/namedtuplelint.py"]                                                                                                       
 ---> Running in 2e3fb61e3a6a                                                                                                                                   
Removing intermediate container 2e3fb61e3a6a                                                                                                                    
 ---> d7207bd418b5                                                                                                                                              
Successfully built d7207bd418b5                                                                                                                                 
Successfully tagged namedtuplelint:latest                                                                                                                       
************* Module namedtuplelint                                                                                                                             
/app/namedtuplelint.py:26:0: E1121: Too many positional arguments for constructor call (too-many-function-args)                                                 
/app/namedtuplelint.py:27:0: E1121: Too many positional arguments for constructor call (too-many-function-args)

**Comment by PCManticore** [comment]
Thanks for the examples @nicktimko !

**Comment by maxis321** [comment]
I'm having a similar issue sunning it on Visual Studio Code. I quite literally have only two parameters being passed, but it drops me this pylint warning.

**Comment by Pierre-Sassoulas** [comment]
This was fixed at some point before 2.14.

**Comment by OJFord** [comment]
1.9.0

https://github.com/PyCQA/pylint/issues/1789#issuecomment-390474813
